feat: Optimize Euler offer status management with unified function

priyanshi-2003 opened the pull request:
- Replaced individual pause_euler_offer, expire_euler_offer, activate_euler_offer functions with unified update_euler_offer_status function
- Created lightweight wrapper functions for backward compatibility
- Maintained existing tool function schemas and mappings
- Improved code maintainability and reduced duplication
- All offer status updates now use single API endpoint: /api/offers/dashboard/{offer_id}/status/update

Dev proof:
https://drive.google.com/file/d/1jp70A97H4nQ7P3uQKim4CZsnF6DII1mu/view?usp=sharing
![Uploading Screenshot 2025-08-25 at 4.47.59 PM.png…]()

badri-singhal commented on app/agents/voice/automatic/tools/juspay/analytics.py:
```diff
@@ -961,10 +1369,30 @@ async def delete_euler_offer(params: FunctionCallParams):
 )
 
+# Helper functions for backward compatibility - all route to update_euler_offer_status
+async def pause_euler_offer(params: FunctionCallParams):
+    """Backward compatibility wrapper for pause functionality"""
+    params.arguments["status"] = "PAUSED"
+    await update_euler_offer_status(params)
+
+async def expire_euler_offer(params: FunctionCallParams):
+    """Backward compatibility wrapper for expire functionality"""
+    params.arguments["status"] = "EXPIRED"
+    await update_euler_offer_status(params)
+
+async def activate_euler_offer(params: FunctionCallParams):
+    """Backward compatibility wrapper for activate functionality"""
+    params.arguments["status"] = "ACTIVE"
+    await update_euler_offer_status(params)
+
```
All these functions are redundant. We only need to keep update_eluer_offer_status, and within its function schema, define the required properties.

priyanshi-2003 replied:
Done

badri-singhal replied:
remove these, as they are not used anywhere

badri-singhal commented on app/agents/voice/automatic/tools/juspay/analytics.py:
```diff
@@ -938,17 +1237,126 @@ async def delete_euler_offer(params: FunctionCallParams):
+    required=["offerCode", "status"]
+)
+
+update_euler_offer_details_function = FunctionSchema(
```
Can you check once, if we can use `create_euler_offer_function` or can we make the existing one to `create_or_update_euler_offer_function`

priyanshi-2003 replied:
Both of them have different endpoints.
Update offer here is not creating new offer and then updating, it will just update the existing one.

> Can you check once, if we can use `create_euler_offer_function` or can we make the existing one to `create_or_update_euler_offer_function`

badri-singhal replied:
I am talking about function schema, not tool executor function

priyanshi-2003 replied:
Changes are done

badri-singhal commented on app/agents/voice/automatic/tools/juspay/analytics.py:
```diff
@@ -712,80 +712,379 @@ async def delete_euler_offer(params: FunctionCallParams):
+            },
+            "status": status or "ACTIVE",
+            "start_time": start_date_iso or "2025-08-25T18:30:00.000Z",
+            "end_time": end_date_iso or "2025-09-26T18:29:59.000Z",
```
If `start_date_iso` or `end_date_iso` is None, why not fail the operation instead of defaulting to arbitrary dates?

priyanshi-2003 replied:
Now fetching all the details from the existing offer details, removed this fallback

badri-singhal commented on app/agents/voice/automatic/tools/juspay/analytics.py:
```diff
@@ -712,80 +712,379 @@ async def delete_euler_offer(params: FunctionCallParams):
+                "payment_channel": [],
+                "benefits": [
+                    {
+                        "type": offer_type or "DISCOUNT",
+                        "calculation_rule": calculation_type or "ABSOLUTE",
+                        "value": discount_value or 10,
```
Same with these fields

badri-singhal commented on app/agents/voice/automatic/tools/juspay/analytics.py:
```diff
@@ -975,5 +1403,9 @@ async def delete_euler_offer(params: FunctionCallParams):
+    "expire_euler_offer": expire_euler_offer,
+    "activate_euler_offer": activate_euler_offer,
+    "update_euler_offer_status": update_euler_offer_status,
+    "update_euler_offer_details": update_euler_offer_details,
```
Also, we can update status using `update_euler_offer_details`. Check once if we need to require both or discuss with @swaroopvarma1 .

priyanshi-2003 replied:
Sure

swaroopvarma1 commented on app/agents/voice/automatic/tools/juspay/analytics.py:
```diff
@@ -938,17 +1347,48 @@ async def delete_euler_offer(params: FunctionCallParams):
-# )
+update_euler_offer_function = FunctionSchema(
+    name="update_euler_offer",
+    description="Intelligently updates offer details or status in the Euler platform. Automatically detects intent based on provided parameters: if only 'status' is provided, updates offer status using optimized endpoint; if other details are provided, performs comprehensive update. This unified function handles status changes (PAUSED/EXPIRED/ACTIVE) and limited modifications. IMPORTANT: The foll ...
```
status' is provided, updates offer status using optimized endpoint -> this part it is not needed for LLM to know it should be generic on what it can do, also remove The following fields are IMMUTABLE and cannot be updated

priyanshi-2003 committed:
feat: Optimize Euler offer management functions

- Modified find_offer_by_code() to return complete offer details instead of just offer_id
- Updated delete_euler_offer() to extract offer_id from returned offer data
- Updated update_euler_offer_status() to extract offer_id from returned offer data
- Updated update_euler_offer_details() to use optimized find_offer_by_code function
- Eliminated redundant API calls by reusing offer data across functions
- Added proper error handling for missing offer_id in all functions
